### PR TITLE
chestnut-headphones: Add onboarding banner to homepage and profile page

### PR DESCRIPTION
**Pull request by taravancil**
## Links
* [chestnut-headphones.glitch.me/](https://chestnut-headphones.glitch.me/)
* [Onboarding design spec](https://docs.google.com/document/d/1iiJnU8COhu0EreV9i5xHaBMJRgeclaxLw6liBz-CNqU/)

welcome

## GIF/Screenshots:
<table>
<tr>
     <td><strong>large viewport</strong></td>
     <td>
  </td>
  </tr>
<tr>
     <td><strong>medium viewport</strong></td>
     <td>
  </td>
  </tr>
  <tr>
     <td><strong>small viewport</strong></td>
     <td></td>
  </tr>
</table>

## Changes:
* Add OnboardingBanner component
* Add CategoriesGrid component (used in OnboardingBanner and /create)
* Move shared constants from /create and the homepage to constants.styl
* Move NewProjectPop out of /header
* Update NewProjectPop to accept props for button styles and button text
* Update NewProjectPop analytics to send user id and the source of the clicks, i.e., navbar, onboarding banner, etc.

## How To Test:

Log in with a test account that doesn't have any projects
- You should see the onboarding banner on the homepage and on your profile page
- Change your screen width, the banner should look good
- Use the accessibility tools to inspect the welcome illustration. The label should be "Welcome"
- Click the "Create New Project" button, it should work.
- Click all the links in the onboarding banner, they should work.

Create a project
- You should not see the onboarding banner

Log out
- You should not see the onboarding banner

Various things to check
- The categories grid on /create still looks good
- The "New Project" button in the navbar still works

## Feedback I'm looking for:
- @ttseng does the banner look good to you?
- @sarahzinger does the data I'm sending along in the new project pop look good?

## Before deploying
- [ ] Run accessibility insights

### CODE REVIEW

**Comment by ttseng** [comment]
nice!  the responsiveness works really well.  

the only change I'd suggest is adding a `margin-bottom: 7px` to `categoriesGridItem wrapItems img` so the category name isn't as close to the image:

image

File: src/components/categories-grid/index.js [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    path: 'learn-to-code',
+    icon: `${CDN_URL}/50f784d9-9995-4fa4-a185-b4b1ea6e77c0%2Flearn.svg`,
+  },
+];
+
+function CategoriesGrid({ categories, wrapItems, className }) {
+  const categoriesToRender = useMemo(
+    () => (categories !== allCategories ? allCategories.filter((category) => categories.includes(category.path)) : allCategories),
+    [categories],
+  );
```
Comment (sarahzinger) [inline src/components/categories-grid/index.js]: what does this do?

File: src/components/categories-grid/index.js [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+  const itemClassNames = cx({
+    categoriesGridItem: true,
+    wrapItems,
+  });
+
+  return (
+    <ul className={classNames(styles.categoriesGrid, className)}>
+      {categoriesToRender.map((category) => (
+        <li key={category.path} className={itemClassNames} style={{ '--bg-color': category.color }}>
```
Comment (sarahzinger) [inline src/components/categories-grid/index.js]: why `--bg-color` and not background-color?

File: src/components/constants.styl [hunk 3]
```diff
@@ -44,7 +45,27 @@ mediumSmallViewport = 592px
 mediumViewport = 670px
 mediumLargeViewport = 800px
 largeViewport = 900px
-hugeViewport = 1080px
+hugeViewport = 1200px
 
 side-padding = 20px
 
+u0 = 2px
```
Comment (sarahzinger) [inline src/components/constants.styl]: so i recognize a lot of this is just moving around, but what is all this? can we put a comment for it? like what does u and t stand for?

File: src/components/new-project-pop/index.js [hunk 4]
```diff
@@ -93,16 +95,30 @@ const useNewProjectAPI = createAPIHook(async (api) => {
...
+  const onOpen = useTracker('open new-project pop', {
+    userId: currentUser.id || null,
+    source,
+  });
 
   return (
-    <PopoverWithButton onOpen={onOpen} buttonProps={{ size: 'small' }} buttonText="New Project">
-      {() => <NewProjectPop projects={projects} />}
+    <PopoverWithButton onOpen={onOpen} buttonProps={{ size: 'small', type: buttonType }} buttonText={buttonText}>
+      {() => <NewProjectPop projects={projects} align={align} analyticsProps={{ userId: currentUser.id || null, source }} />}
```
Comment (sarahzinger) [inline src/components/new-project-pop/index.js]: so this is totes chill but something we can do besides passing along props like this is to use AnalyticsContext instead if you're interested:
https://github.com/FogCreek/Glitch-Community/blob/4497ae553e1b37e3fee4268581ab90ec80329e1b/src/state/segment-analytics.js#L18
so like if this is on the user page for example you could pass currentUser.id at a higher level here and any analytics call on the userpage will have that 
https://github.com/FogCreek/Glitch-Community/blob/97eec7d4339492f103fab8f19cea8e8f31782ff1/src/presenters/pages/user.js#L217
and if you need to build on that origin in the function where you track things itself you can follow this inherited thing we do here:
https://github.com/FogCreek/Glitch-Community/blob/d5675ed420651b05f2c0dc13b2e46cd6b504095a/src/components/collection/create-collection-pop.js#L160

File: src/components/onboarding-banner/illustration.js [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+import styles from './illustration.styl';
+
+const Illustration = () => (
+  <svg viewBox="0 0 332 263" role="img">
+    <title>Welcome</title>
+    <desc>A computer with a hand giving a welcoming wave</desc>
+    <defs>
+      <path
+        d="M4.233.876c-2.53.959-3.958 3.376-3.192 5.402C1.81 8.3 4.482 9.164 7.013 8.204c2.527-.959 3.956-3.377 3.19-5.4C9.646 1.34 8.095.483 6.324.483a5.91 5.91 0 0 0-2.092.392z"
+        id="path-1"
```
Comment (sarahzinger) [inline src/components/onboarding-banner/illustration.js]: i think we want to remove all ids from our inline svgs just incase they are ever reused on a page.

File: src/components/onboarding-banner/index.js [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    ? { backgroundImage: 'url(https://cdn.glitch.com/b065beeb-4c71-4a9c-a8aa-4548e266471f%2Fuser-pattern.svg)', backgroundColor: lightColors[currentUser.id % 4] } : null;
+  const sourceLabel = `${isHomepage ? 'homepage' : 'profile page'} onboarding banner`;
+
+  return (
+    <div className={styles.banner} style={backgroundStyles}>
+      <div className={styles.illustration}>
+        <Illustration />
+      </div>
+
+      <div className={styles.actions} style={{ background: isHomepage ? 'rgba(255, 255, 255, .875)' : 'none' }}>
```
Comment (sarahzinger) [inline src/components/onboarding-banner/index.js]: why not do this with css?

File: src/components/onboarding-banner/illustration.styl [hunk 7]
```diff
@@ -0,0 +1,7 @@
+.hand
+  animation: wave cubic-bezier(0.24, 0.45, 1, 1) 2s infinite forwards
+  transform-origin: 60px 121px
+
+@keyframes wave
+  33%
+    transform: rotate(-7deg);
```
Comment (sarahzinger) [inline src/components/onboarding-banner/illustration.styl]: I think you'll want to add a media query to not animate if the user prefers no motion https://github.com/FogCreek/Glitch-Community/blob/b33e7f7909495f496c89e34735de5693294eb8cc/src/presenters/pages/home-v2/animations.styl#L346

File: src/components/new-project-pop/index.js [hunk 8]
```diff
@@ -93,16 +95,30 @@ const useNewProjectAPI = createAPIHook(async (api) => {
...
     </PopoverWithButton>
   );
 }
 
+NewProjectPopButton.propTypes = {
```
Comment (clottman) [inline src/components/new-project-pop/index.js]: Looks like we're missing prop types here - buttonType and source.

File: src/components/onboarding-banner/illustration.styl [hunk 9]
```diff
@@ -0,0 +1,7 @@
+.hand
+  animation: wave cubic-bezier(0.24, 0.45, 1, 1) 2s infinite forwards
+  transform-origin: 60px 121px
+
+@keyframes wave
```
Comment (clottman) [inline src/components/onboarding-banner/illustration.styl]: Can we add in a media query to honor the user's `prefers-reduced-motion` setting, if set?

File: src/components/new-project-pop/index.js (hunk 4 above)
Comment (sarahzinger) [reply src/components/new-project-pop/index.js]: oh just realized you're also using https://github.com/FogCreek/Glitch-Community/blob/4497ae553e1b37e3fee4268581ab90ec80329e1b/src/state/segment-analytics.js#L68 but I think that will work the same way, we can double check together using segment if you want!

File: src/components/categories-grid/index.js (hunk 1 above)
Comment (taravancil) [reply src/components/categories-grid/index.js]: This lets us optionally render only *some* of the categories. There are 8 total, but for example, the [grid in the onboarding banner](https://github.com/FogCreek/Glitch-Community/pull/836/files#diff-ad7cb9c5b67e9f30910fcf625f867aa1R59) only shows 6/8: 'games', 'music', 'art', 'handy-bots', 'learn-to-code', 'tools-for-work'

File: src/components/categories-grid/index.js (hunk 2 above)
Comment (taravancil) [reply src/components/categories-grid/index.js]: 🤷‍♀ idk tbh, you prefer `--background-color`?

File: src/components/constants.styl (hunk 3 above)
Comment (taravancil) [reply src/components/constants.styl]: I honestly do not know. @modernserf introduced these with the new homepage

File: src/components/categories-grid/index.js (hunk 1 above)
Comment (sarahzinger) [reply src/components/categories-grid/index.js]: cooool

File: src/components/categories-grid/index.js (hunk 2 above)
Comment (sarahzinger) [reply src/components/categories-grid/index.js]: i guess why does it have the `--` before it? is it deprecated? I thought you could do `style={{backgroundColor: category.color}}`

but maybe it doesn't matter 🤷‍♀

Comment (taravancil) [reply src/components/categories-grid/index.js]: Oh I see, I'm setting a CSS variable - that's why it has the `--` in front. I could do it directly inline, but I'm not sure it matters that much

File: src/components/onboarding-banner/index.js (hunk 6 above)
Comment (taravancil) [reply src/components/onboarding-banner/index.js]: No good reason really, other than I think I wrote this at 2am 🤦🏻‍♀️

File: src/components/new-project-pop/index.js (hunk 8 above)
Comment (taravancil) [reply src/components/new-project-pop/index.js]: Added `source`, but invalid `buttonType`s will get caught in `Button` itself. Do you think it's worth duplicating for the sake of documentation?

File: src/components/new-project-pop/index.js (hunk 4 above)
Comment (taravancil) [reply src/components/new-project-pop/index.js]: Oh sick, I'm making these updates now. Maybe if we have time later we can double check everything in Segment

File: src/components/constants.styl (hunk 3 above)
Comment (modernserf) [reply src/components/constants.styl]: The "u" units are margin & padding units; the "t"s are font sizes. Those all have clearer names in the shared component library

**Comment by taravancil** [comment]
This is waiting of the infra team to whitelist the `ab-tests` cookie. They have a PR open, so I should be able to pick this back up after the holiday.

**Comment by ttseng** [comment]
@keithk re: the links on the create page, are you referring to the categories at the bottom (rather than the framework or platform starters)?

image

(also, is that a concern for a separate ticket)?

**Comment by taravancil** [comment]
A few notes on the changes I've pushed since this was last reviewed:

**A/B testing**
A/B testing is implemented! We're using a 50/50 onboard/don't onboard split. Pretty simple.

- [shared/ab-tests.js](https://github.com/FogCreek/Glitch-Community/pull/836/files#diff-58dbba706ff4b69f9bbd8ce359f855ea)
- [onboarding-banner/index.js#L24](https://github.com/FogCreek/Glitch-Community/pull/836/files#diff-ad7cb9c5b67e9f30910fcf625f867aa1R24)

**AnalyticsContext**
Per Sarah's suggestion, I moved all of the custom analytics props into `<AnalyticsContext>`

- [onboarding-banner/index.js#L53](https://github.com/FogCreek/Glitch-Community/pull/836/files#diff-ad7cb9c5b67e9f30910fcf625f867aa1R53)
- [header/index.js#L34](https://github.com/FogCreek/Glitch-Community/pull/836/files#diff-5c2ad23269ed1bd2a33c79c14b77e315R34)

**Shared components**
I pulled latest, which pulls in a few shared components. I updated the one emoji in the onboarding banner to use the `<Icon>` component. The shared-components library doesn't include `backhandIndex`, so until [this PR](https://github.com/FogCreek/shared-components/pull/45/files) is deployed, we're blocked

- [onboarding-banner/index.js#L85](https://github.com/FogCreek/Glitch-Community/pull/836/files#diff-ad7cb9c5b67e9f30910fcf625f867aa1R85)

**Comment by taravancil** [comment]
Here's what a click on the "New Project" button in the onboarding banner will look like in Segment. It gives you `properties.origin`, which indicates whether the "New Project" button was clicked in the header or the onboarding banner.

Anything else we'll need to include?

```
{
  "anonymousId": "1234",
  "context": {
    "groupId": "0",
    "ip": "172.110.88.61",
    "library": {
      "name": "analytics.js",
      "version": "3.9.0"
    },
    "page": {
      "path": "/",
      "referrer": "",
      "search": "",
      "title": "Glitch",
      "url": "https://chestnut-headphones.glitch.me/"
    },
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.14; rv:70.0) Gecko/20100101 Firefox/70.0"
  },
  "event": "open new-project pop",
  "integrations": {},
  "messageId": "ajs-e73772492f6bfeb946c0df7059517eb1",
  "originalTimestamp": "2019-09-10T14:38:49.258Z",
  "properties": {
    "origin": "homepage onboarding banner"
  },
  "receivedAt": "2019-09-10T14:38:49.305Z",
  "sentAt": "2019-09-10T14:38:49.260Z",
  "timestamp": "2019-09-10T14:38:49.303Z",
  "type": "track",
  "userId": "aUserId"
}
```

**Comment by sarahzinger** [comment]
i'm currently seeing the onboarding banner on my user page but not on the homepage, is that a desired state?

**Comment by GregWeil** [comment]
psa: use `document.cookie = 'ab-tests='+JSON.stringify({...AB_TESTS, Onboarding:'banner'})` to force yourself into the right a/b testing group

**Comment by taravancil** [comment]
> i'm currently seeing the onboarding banner on my user page but not on the homepage

🤦‍♀ lmao no, i had toggled a flag for testing and never switched it back. @sarahzinger should be fixed now

Also note that you need to use a test account that has no projects to see the onboarding banner

**Comment by ttseng** [comment]
Is it just me or the CTA styling for the `Create New Project` button not being applied?

image

(should be teal).

Otherwise, it looks great!

**Comment by taravancil** [comment]
Yep, thanks for the catch Tiff! Must've gotten lost when I pulled in the styled components changes.

**Comment by ttseng** [comment]
wave

Fixed hand direction + adjusted animation for computer illustration

**Comment by taravancil** [comment]
Bumped the shared-compononents lib and added a fix to prevent the onboarding banner from being rendered on the server for users who shouldn't see it

**Comment by sarahzinger** [comment]
still seems good to me :)

**Pull request merged**